# WELCOME TO MY NORTHWIND SQL PROJECT
---
This project shows how to use SQL to analyze data to answer the business questions by the management team. I connected the Northwind postgreSQL database to Jupyter to enable documentation during coding

## Connecting PostgreSQL database of Northwind Project to Jupyter
--- 

In [1]:
!pip install ipython-sql
!pip install sqlalchemy
!pip install psycopg2
%load_ext sql
from sqlalchemy import create_engine
%sql postgresql://postgres:@localhost:5432/northwind
engine=create_engine('postgresql://postgres:@localhost:5432/northwind')

# Checking the schema from the Northwind database
---

In [2]:
%%sql

SELECT table_name AS name,
       table_type AS type
  FROM information_schema.tables
 WHERE table_schema = 'public' AND table_type IN ('BASE TABLE', 'VIEW');

 * postgresql://postgres:***@localhost:5432/northwind
14 rows affected.


[('territories', 'BASE TABLE'),
 ('order_details', 'BASE TABLE'),
 ('employee_territories', 'BASE TABLE'),
 ('us_states', 'BASE TABLE'),
 ('customers', 'BASE TABLE'),
 ('orders', 'BASE TABLE'),
 ('employees', 'BASE TABLE'),
 ('shippers', 'BASE TABLE'),
 ('products', 'BASE TABLE'),
 ('categories', 'BASE TABLE'),
 ('suppliers', 'BASE TABLE'),
 ('region', 'BASE TABLE'),
 ('customer_demographics', 'BASE TABLE'),
 ('customer_customer_demo', 'BASE TABLE')]

![Image](https://github.com/pthom/northwind_psql/blob/master/ER.png?raw=true)

[Reference](https://github.com/pthom/northwind_psql/blob/master/ER.png)

# In order to understand the data, we should check the different tables and see how we can relate them to each other. Using queries to check some of the tables and their columns:
---

## customers


In [3]:
%%sql

SELECT *
  FROM customers
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[('ALFKI', 'Alfreds Futterkiste', 'Maria Anders', 'Sales Representative', 'Obere Str. 57', 'Berlin', None, '12209', 'Germany', '030-0074321', '030-0076545'),
 ('ANATR', 'Ana Trujillo Emparedados y helados', 'Ana Trujillo', 'Owner', 'Avda. de la Constitución 2222', 'México D.F.', None, '05021', 'Mexico', '(5) 555-4729', '(5) 555-3745'),
 ('ANTON', 'Antonio Moreno Taquería', 'Antonio Moreno', 'Owner', 'Mataderos  2312', 'México D.F.', None, '05023', 'Mexico', '(5) 555-3932', None),
 ('AROUT', 'Around the Horn', 'Thomas Hardy', 'Sales Representative', '120 Hanover Sq.', 'London', None, 'WA1 1DP', 'UK', '(171) 555-7788', '(171) 555-6750'),
 ('BERGS', 'Berglunds snabbköp', 'Christina Berglund', 'Order Administrator', 'Berguvsvägen  8', 'Luleå', None, 'S-958 22', 'Sweden', '0921-12 34 65', '0921-12 34 67')]

## orders

In [4]:
%%sql

SELECT *
  FROM orders
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil'),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France'),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium')]

## products

In [5]:
%%sql

SELECT *
  FROM products
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(1, 'Chai', 8, 1, '10 boxes x 30 bags', 18.0, 39, 0, 10, 1),
 (2, 'Chang', 1, 1, '24 - 12 oz bottles', 19.0, 17, 40, 25, 1),
 (3, 'Aniseed Syrup', 1, 2, '12 - 550 ml bottles', 10.0, 13, 70, 25, 0),
 (4, "Chef Anton's Cajun Seasoning", 2, 2, '48 - 6 oz jars', 22.0, 53, 0, 0, 0),
 (5, "Chef Anton's Gumbo Mix", 2, 2, '36 boxes', 21.35, 0, 0, 0, 1)]

## order_details

In [7]:
%%sql

SELECT *
  FROM order_details
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 11, 14.0, 12, 0.0),
 (10248, 42, 9.8, 10, 0.0),
 (10248, 72, 34.8, 5, 0.0),
 (10249, 14, 18.6, 9, 0.0),
 (10249, 51, 42.4, 40, 0.0)]

# BUSINESS QUESTIONS:

## 1. In order to recognize their performance, the management wants to determine who are the top-performing employees in the organization

Using `JOIN` to combine orders, employees, and order_details tables to get the top performing employees as `CTE` based on their total sales and rank them using the `RANK()` function:

In [8]:
%%sql

WITH top_sales_table AS (
    
SELECT e.employee_id AS "Employee ID",
       e.last_name AS "Last Name",
       e.first_name AS "First Name",
       ROUND(CAST(SUM(quantity*unit_price*(1-discount)) AS numeric),2) AS "Total Sales"
  FROM orders AS o
  JOIN employees AS e
    ON o.employee_id = e.employee_id
  JOIN order_details AS od
    ON o.order_id = od.order_id
 GROUP BY e.employee_id
 
)

SELECT *,RANK() OVER(
         ORDER BY "Total Sales" DESC
         ) AS "Employee Rank"
  FROM top_sales_table;


 * postgresql://postgres:***@localhost:5432/northwind
9 rows affected.


[(4, 'Peacock', 'Margaret', Decimal('232890.85'), 1),
 (3, 'Leverling', 'Janet', Decimal('202812.84'), 2),
 (1, 'Davolio', 'Nancy', Decimal('192107.60'), 3),
 (2, 'Fuller', 'Andrew', Decimal('166537.76'), 4),
 (8, 'Callahan', 'Laura', Decimal('126862.28'), 5),
 (7, 'King', 'Robert', Decimal('124568.23'), 6),
 (9, 'Dodsworth', 'Anne', Decimal('77308.07'), 7),
 (6, 'Suyama', 'Michael', Decimal('73913.13'), 8),
 (5, 'Buchanan', 'Steven', Decimal('68792.28'), 9)]

Based from the result of our query, Margaret Peacock is the top performing employee with a whopping 232,890.85 USD total sales. On the other hand, Steven Buchanan is at the bottom with 68,792.28 USD total sales. He will be one of the management's targets, along with the other low performing employees,for upskilling and trainings to help them improve their work.

## 2. What is the running total of the monthly sales?

Using `CTE` for the monthly sales and then using `SUM() OVER()` to calculate the running sales each month:

In [9]:
%%sql

WITH monthly_sales_table AS (
SELECT DATE_TRUNC('month',o.order_date)::DATE AS month_date,
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS total_sales
  FROM order_details AS od
  JOIN orders AS o
 USING (order_id)
 GROUP BY DATE_TRUNC('month',o.order_date) 
 ORDER BY DATE_TRUNC('month',o.order_date) 
)

SELECT *, 
       SUM(total_sales) OVER(
       ORDER BY month_date   
       ) AS monthly_running_sales
  FROM monthly_sales_table

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(datetime.date(1996, 7, 1), Decimal('27861.90'), Decimal('27861.90')),
 (datetime.date(1996, 8, 1), Decimal('25485.28'), Decimal('53347.18')),
 (datetime.date(1996, 9, 1), Decimal('26381.40'), Decimal('79728.58')),
 (datetime.date(1996, 10, 1), Decimal('37515.72'), Decimal('117244.30')),
 (datetime.date(1996, 11, 1), Decimal('45600.05'), Decimal('162844.35')),
 (datetime.date(1996, 12, 1), Decimal('45239.63'), Decimal('208083.98')),
 (datetime.date(1997, 1, 1), Decimal('61258.07'), Decimal('269342.05')),
 (datetime.date(1997, 2, 1), Decimal('38483.63'), Decimal('307825.68')),
 (datetime.date(1997, 3, 1), Decimal('38547.22'), Decimal('346372.90')),
 (datetime.date(1997, 4, 1), Decimal('53032.95'), Decimal('399405.85')),
 (datetime.date(1997, 5, 1), Decimal('53781.29'), Decimal('453187.14')),
 (datetime.date(1997, 6, 1), Decimal('36362.80'), Decimal('489549.94')),
 (datetime.date(1997, 7, 1), Decimal('51020.86'), Decimal('540570.80')),
 (datetime.date(1997, 8, 1), Decimal('47287.67'), Decimal('587858.47')),
 (datetime.date(1997, 9, 1), Decimal('55629.24'), Decimal('643487.71')),
 (datetime.date(1997, 10, 1), Decimal('66749.23'), Decimal('710236.94')),
 (datetime.date(1997, 11, 1), Decimal('43533.81'), Decimal('753770.75')),
 (datetime.date(1997, 12, 1), Decimal('71398.43'), Decimal('825169.18')),
 (datetime.date(1998, 1, 1), Decimal('94222.11'), Decimal('919391.29')),
 (datetime.date(1998, 2, 1), Decimal('99415.29'), Decimal('1018806.58')),
 (datetime.date(1998, 3, 1), Decimal('104854.16'), Decimal('1123660.74')),
 (datetime.date(1998, 4, 1), Decimal('123798.68'), Decimal('1247459.42')),
 (datetime.date(1998, 5, 1), Decimal('18333.63'), Decimal('1265793.05'))]

## 3. What is the monthly growth for each month?

Using `DATE_TRUNC()` to extract the dates, we can check the monthly growth by using`LAG()` to check the previous row's sales and compare it with the current row's sales value:

In [10]:
%%sql

WITH monthly_sales_table AS (
SELECT DATE_TRUNC('month',o.order_date)::DATE AS month_date,
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS total_sales
  FROM order_details AS od
  JOIN orders AS o
 USING (order_id)
 GROUP BY DATE_TRUNC('month',o.order_date) 
 ORDER BY DATE_TRUNC('month',o.order_date) 
),

previous_sales_table AS (
SELECT *,
       LAG(total_sales) OVER (
       ORDER BY month_date
       ) AS previous_month_sales
  FROM monthly_sales_table
    
)

SELECT EXTRACT(year FROM month_date) AS year,
       EXTRACT(month FROM month_date) AS month,
       ROUND(((total_sales-previous_month_sales)/previous_month_sales)*100,2) AS growth_rate
  FROM previous_sales_table

    

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(Decimal('1996'), Decimal('7'), None),
 (Decimal('1996'), Decimal('8'), Decimal('-8.53')),
 (Decimal('1996'), Decimal('9'), Decimal('3.52')),
 (Decimal('1996'), Decimal('10'), Decimal('42.21')),
 (Decimal('1996'), Decimal('11'), Decimal('21.55')),
 (Decimal('1996'), Decimal('12'), Decimal('-0.79')),
 (Decimal('1997'), Decimal('1'), Decimal('35.41')),
 (Decimal('1997'), Decimal('2'), Decimal('-37.18')),
 (Decimal('1997'), Decimal('3'), Decimal('0.17')),
 (Decimal('1997'), Decimal('4'), Decimal('37.58')),
 (Decimal('1997'), Decimal('5'), Decimal('1.41')),
 (Decimal('1997'), Decimal('6'), Decimal('-32.39')),
 (Decimal('1997'), Decimal('7'), Decimal('40.31')),
 (Decimal('1997'), Decimal('8'), Decimal('-7.32')),
 (Decimal('1997'), Decimal('9'), Decimal('17.64')),
 (Decimal('1997'), Decimal('10'), Decimal('19.99')),
 (Decimal('1997'), Decimal('11'), Decimal('-34.78')),
 (Decimal('1997'), Decimal('12'), Decimal('64.01')),
 (Decimal('1998'), Decimal('1'), Decimal('31.97')),
 (Decimal('1998'), Decimal('2'), Decimal('5.51')),
 (Decimal('1998'), Decimal('3'), Decimal('5.47')),
 (Decimal('1998'), Decimal('4'), Decimal('18.07')),
 (Decimal('1998'), Decimal('5'), Decimal('-85.19'))]

Based from the growth table, we can see that the peak of sales growth was on December of 1997 at 64.01%

## 4. In order to increase their sales even further, the management asks to identify the top customers who are buying in bulk or purchasing high-end products

Using `CTE` to check for the customer's sales and then using `CASE` to label those sales that are Above Average, Average, or Below Average:

In [11]:
%%sql

WITH customer_sales_table AS (
SELECT o.customer_id, o.order_id,
       SUM(od.quantity*od.unit_price*(1-od.discount)) AS customer_sales
  FROM orders AS o
  JOIN order_details AS od
 USING (order_id)
 GROUP BY o.customer_id, o.order_id
 ORDER BY o.customer_id
)

SELECT customer_id,order_id,customer_sales,
       CASE
           WHEN customer_sales > AVG(customer_sales) OVER () THEN 'Above Average'
           WHEN customer_sales = AVG(customer_sales) OVER () THEN 'Average'
           ELSE 'Below Average'
        END AS sales_rating
  FROM customer_sales_table;



 * postgresql://postgres:***@localhost:5432/northwind
830 rows affected.


[('ALFKI', 10643, 814.4999828338623, 'Below Average'),
 ('ALFKI', 10692, 878.0000305175781, 'Below Average'),
 ('ALFKI', 10702, 330.0, 'Below Average'),
 ('ALFKI', 10835, 845.799999922514, 'Below Average'),
 ('ALFKI', 10952, 471.19999665021896, 'Below Average'),
 ('ALFKI', 11011, 933.4999996051192, 'Below Average'),
 ('ANATR', 10625, 479.75, 'Below Average'),
 ('ANATR', 10926, 514.3999910354614, 'Below Average'),
 ('ANATR', 10759, 320.0, 'Below Average'),
 ('ANATR', 10308, 88.79999923706055, 'Below Average'),
 ('ANTON', 10507, 749.0624947473407, 'Below Average'),
 ('ANTON', 10365, 403.1999816894531, 'Below Average'),
 ('ANTON', 10573, 2081.9999809265137, 'Above Average'),
 ('ANTON', 10535, 1940.8499933533371, 'Above Average'),
 ('ANTON', 10682, 375.5, 'Below Average'),
 ('ANTON', 10677, 813.364982623458, 'Below Average'),
 ('ANTON', 10856, 660.0, 'Below Average'),
 ('AROUT', 10743, 319.1999997496605, 'Below Average'),
 ('AROUT', 10558, 2142.89998626709, 'Above Average'),
 ('AROUT', 10953, 4441.249996516854, 'Above Average'),
 ('AROUT', 10453, 407.69999675005675, 'Below Average'),
 ('AROUT', 10768, 1477.0, 'Below Average'),
 ('AROUT', 10383, 898.9999961853027, 'Below Average'),
 ('AROUT', 10793, 191.09999465942383, 'Below Average'),
 ('AROUT', 11016, 491.5, 'Below Average'),
 ('AROUT', 10920, 390.0, 'Below Average'),
 ('AROUT', 10355, 480.0000071525574, 'Below Average'),
 ('AROUT', 10707, 1640.999997496605, 'Above Average'),
 ('AROUT', 10864, 282.0, 'Below Average'),
 ('AROUT', 10741, 227.9999991506338, 'Below Average'),
 ('BERGS', 10280, 613.200008392334, 'Below Average'),
 ('BERGS', 10733, 1458.9999694824219, 'Below Average'),
 ('BERGS', 10654, 601.829997973591, 'Below Average'),
 ('BERGS', 10445, 174.89999914169312, 'Below Average'),
 ('BERGS', 10837, 1064.4999904632568, 'Below Average'),
 ('BERGS', 10572, 1501.0849882041662, 'Below Average'),
 ('BERGS', 10857, 2048.2124948501587, 'Above Average'),
 ('BERGS', 10672, 3815.249994110316, 'Above Average'),
 ('BERGS', 10866, 1096.1999816894531, 'Below Average'),
 ('BERGS', 10778, 96.49999618530273, 'Below Average'),
 ('BERGS', 10924, 1835.6999558389189, 'Above Average'),
 ('BERGS', 10524, 3192.6499910354614, 'Above Average'),
 ('BERGS', 10384, 2222.4000968933105, 'Above Average'),
 ('BERGS', 10689, 472.5, 'Below Average'),
 ('BERGS', 10278, 1488.7999877929688, 'Below Average'),
 ('BERGS', 10444, 1031.699993133545, 'Below Average'),
 ('BERGS', 10875, 709.5499949343503, 'Below Average'),
 ('BERGS', 10626, 1503.5999908447266, 'Below Average'),
 ('BLAUS', 10956, 677.0, 'Below Average'),
 ('BLAUS', 11058, 858.0, 'Below Average'),
 ('BLAUS', 10853, 625.0, 'Below Average'),
 ('BLAUS', 10501, 148.99999618530273, 'Below Average'),
 ('BLAUS', 10614, 464.0, 'Below Average'),
 ('BLAUS', 10509, 136.79999542236328, 'Below Average'),
 ('BLAUS', 10582, 330.0, 'Below Average'),
 ('BLONP', 10265, 1176.0000228881836, 'Below Average'),
 ('BLONP', 10826, 730.0, 'Below Average'),
 ('BLONP', 10584, 593.7499995343387, 'Below Average'),
 ('BLONP', 10628, 450.0, 'Below Average'),
 ('BLONP', 10566, 1760.999988913536, 'Above Average'),
 ('BLONP', 10559, 520.4099952431023, 'Below Average'),
 ('BLONP', 10449, 1838.2000408172607, 'Above Average'),
 ('BLONP', 10360, 7390.200078964233, 'Above Average'),
 ('BLONP', 10436, 1994.5199911314248, 'Above Average'),
 ('BLONP', 10679, 660.0, 'Below Average'),
 ('BLONP', 10297, 1419.9999618530273, 'Below Average'),
 ('BOLID', 10801, 3026.85001373291, 'Above Average'),
 ('BOLID', 10970, 223.99999916553497, 'Below Average'),
 ('BOLID', 10326, 982.0000057220459, 'Below Average'),
 ('BONAP', 10932, 1788.629996739775, 'Above Average'),
 ('BONAP', 10525, 818.3999944388867, 'Below Average'),
 ('BONAP', 11076, 792.7499985694885, 'Below Average'),
 ('BONAP', 10511, 2549.9999821186066, 'Above Average'),
 ('BONAP', 10663, 1930.3999876141547, 'Above Average'),
 ('BONAP', 10730, 484.262503244169, 'Below Average'),
 ('BONAP', 10755, 1948.5, 'Above Average'),
 ('BONAP', 10940, 360.0, 'B

Using the two `CTE` to determine the `COUNT()` of the sales rating that are Above Average

In [12]:
%%sql

WITH customer_sales_table AS (
SELECT o.customer_id, o.order_id,
       SUM(od.quantity*od.unit_price*(1-od.discount)) AS customer_sales
  FROM orders AS o
  JOIN order_details AS od
 USING (order_id)
 GROUP BY o.customer_id, o.order_id
 ORDER BY o.customer_id
),

customer_sales_rating_table AS (
SELECT customer_id,order_id,customer_sales,
       CASE
           WHEN customer_sales > AVG(customer_sales) OVER () THEN 'Above Average'
           WHEN customer_sales = AVG(customer_sales) OVER () THEN 'Average'
           ELSE 'Below Average'
        END AS sales_rating
  FROM customer_sales_table
)

SELECT customer_id,COUNT(sales_rating) AS above_average_orders
  FROM customer_sales_rating_table
 WHERE sales_rating = 'Above Average'
 GROUP BY customer_id
 ORDER BY above_average_orders DESC
 LIMIT 10;

 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[('SAVEA', 26),
 ('ERNSH', 26),
 ('QUICK', 22),
 ('HUNGO', 11),
 ('RATTC', 10),
 ('BONAP', 8),
 ('FOLKO', 8),
 ('RICSU', 7),
 ('HILAA', 7),
 ('FRANK', 7)]

The top three customer_ids with above average orders are SAVEA, ERNSH, and QUICK. It is recommended to target the top 3 customer_ids for promotion, special offers, and marketing to increase their generated sales.

## 5. What is the percentage of total sales for each product category?

Using `JOIN` to check for the total sales for each product id and category:

In [19]:
%%sql


SELECT p.category_id,c.category_name,od.product_id, 
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS product_sales
  FROM order_details AS od
  JOIN products AS p
 USING (product_id)
  JOIN categories AS c
 USING (category_id)
 GROUP BY od.product_id,category_id,category_name
 ORDER BY category_id
 LIMIT 20;


 * postgresql://postgres:***@localhost:5432/northwind
20 rows affected.


[(1, 'Beverages', 1, Decimal('12788.10')),
 (1, 'Beverages', 2, Decimal('16355.96')),
 (1, 'Beverages', 24, Decimal('4504.36')),
 (1, 'Beverages', 34, Decimal('6350.40')),
 (1, 'Beverages', 35, Decimal('13644.00')),
 (1, 'Beverages', 38, Decimal('141396.74')),
 (1, 'Beverages', 39, Decimal('12294.54')),
 (1, 'Beverages', 43, Decimal('23526.70')),
 (1, 'Beverages', 67, Decimal('2396.80')),
 (1, 'Beverages', 70, Decimal('10672.65')),
 (1, 'Beverages', 75, Decimal('8177.49')),
 (1, 'Beverages', 76, Decimal('15760.44')),
 (2, 'Condiments', 3, Decimal('3044.00')),
 (2, 'Condiments', 4, Decimal('8567.90')),
 (2, 'Condiments', 5, Decimal('5347.20')),
 (2, 'Condiments', 6, Decimal('7137.00')),
 (2, 'Condiments', 8, Decimal('12772.00')),
 (2, 'Condiments', 15, Decimal('1784.82')),
 (2, 'Condiments', 44, Decimal('9915.95')),
 (2, 'Condiments', 61, Decimal('14352.60'))]

From the `CTE` above we can calculate the percentage of total sales for each category using `OVER()`

In [14]:
%%sql

WITH product_sales_table AS (
SELECT p.category_id,c.category_name,od.product_id, 
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS product_sales
  FROM order_details AS od
  JOIN products AS p
 USING (product_id)
  JOIN categories AS c
 USING (category_id)
 GROUP BY od.product_id,category_id,category_name
 ORDER BY category_id
)

SELECT category_id,category_name,SUM(product_sales) AS sales_per_category,
       SUM(product_sales)*100/SUM(SUM(product_sales)) OVER () AS percentage_of_total_sales
  FROM product_sales_table
 GROUP BY category_id,category_name
 ORDER BY percentage_of_total_sales DESC;

 * postgresql://postgres:***@localhost:5432/northwind
8 rows affected.


[(1, 'Beverages', Decimal('267868.18'), Decimal('21.1620834951028013')),
 (4, 'Dairy Products', Decimal('234507.28'), Decimal('18.5265104633534721')),
 (3, 'Confections', Decimal('167357.23'), Decimal('13.2215318548441379')),
 (6, 'Meat/Poultry', Decimal('163022.36'), Decimal('12.8790690775168111')),
 (8, 'Seafood', Decimal('131261.73'), Decimal('10.3699203465362710')),
 (2, 'Condiments', Decimal('106047.09'), Decimal('8.3779169776443075')),
 (7, 'Produce', Decimal('99984.58'), Decimal('7.8989674330963299')),
 (5, 'Grains/Cereals', Decimal('95744.59'), Decimal('7.5640003519058692'))]

The Beverages product category has the highest percentage of total sales at 21.16%. It is recommended to prioritize this category for marketing strategies and restocking of supply

## 6. What are the top products of each category?

Using `JOIN` to calculate the sales for each product:

In [20]:
%%sql


SELECT p.category_id,c.category_name,od.product_id,p.product_name,
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS product_sales
  FROM order_details AS od
  JOIN products AS p
 USING (product_id)
  JOIN categories AS c
 USING (category_id)
 GROUP BY od.product_id,p.product_name,p.category_id,c.category_name
 ORDER BY category_id
 LIMIT 20;


 * postgresql://postgres:***@localhost:5432/northwind
20 rows affected.


[(1, 'Beverages', 1, 'Chai', Decimal('12788.10')),
 (1, 'Beverages', 2, 'Chang', Decimal('16355.96')),
 (1, 'Beverages', 24, 'Guaraná Fantástica', Decimal('4504.36')),
 (1, 'Beverages', 34, 'Sasquatch Ale', Decimal('6350.40')),
 (1, 'Beverages', 35, 'Steeleye Stout', Decimal('13644.00')),
 (1, 'Beverages', 38, 'Côte de Blaye', Decimal('141396.74')),
 (1, 'Beverages', 39, 'Chartreuse verte', Decimal('12294.54')),
 (1, 'Beverages', 43, 'Ipoh Coffee', Decimal('23526.70')),
 (1, 'Beverages', 67, 'Laughing Lumberjack Lager', Decimal('2396.80')),
 (1, 'Beverages', 70, 'Outback Lager', Decimal('10672.65')),
 (1, 'Beverages', 75, 'Rhönbräu Klosterbier', Decimal('8177.49')),
 (1, 'Beverages', 76, 'Lakkalikööri', Decimal('15760.44')),
 (2, 'Condiments', 3, 'Aniseed Syrup', Decimal('3044.00')),
 (2, 'Condiments', 4, "Chef Anton's Cajun Seasoning", Decimal('8567.90')),
 (2, 'Condiments', 5, "Chef Anton's Gumbo Mix", Decimal('5347.20')),
 (2, 'Condiments', 6, "Grandma's Boysenberry Spread", Decimal('7137.00')),
 (2, 'Condiments', 8, 'Northwoods Cranberry Sauce', Decimal('12772.00')),
 (2, 'Condiments', 15, 'Genen Shouyu', Decimal('1784.82')),
 (2, 'Condiments', 44, 'Gula Malacca', Decimal('9915.95')),
 (2, 'Condiments', 61, "Sirop d'érable", Decimal('14352.60'))]

Using the `CTE` above and `DENSE_RANK()` to rank the product sales from highest to lowest for each product category:

In [21]:
%%sql

WITH top_products_table AS (
SELECT p.category_id,c.category_name,od.product_id,p.product_name,
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS product_sales
  FROM order_details AS od
  JOIN products AS p
 USING (product_id)
  JOIN categories AS c
 USING (category_id)
 GROUP BY od.product_id,p.product_name,p.category_id,c.category_name
 ORDER BY category_id
    
)

SELECT category_id,product_id,product_name,product_sales, 
       DENSE_RANK() OVER (
       PARTITION BY category_id
       ORDER BY product_sales DESC
       ) AS product_ranking
  FROM top_products_table
 LIMIT 20;





 * postgresql://postgres:***@localhost:5432/northwind
20 rows affected.


[(1, 38, 'Côte de Blaye', Decimal('141396.74'), 1),
 (1, 43, 'Ipoh Coffee', Decimal('23526.70'), 2),
 (1, 2, 'Chang', Decimal('16355.96'), 3),
 (1, 76, 'Lakkalikööri', Decimal('15760.44'), 4),
 (1, 35, 'Steeleye Stout', Decimal('13644.00'), 5),
 (1, 1, 'Chai', Decimal('12788.10'), 6),
 (1, 39, 'Chartreuse verte', Decimal('12294.54'), 7),
 (1, 70, 'Outback Lager', Decimal('10672.65'), 8),
 (1, 75, 'Rhönbräu Klosterbier', Decimal('8177.49'), 9),
 (1, 34, 'Sasquatch Ale', Decimal('6350.40'), 10),
 (1, 24, 'Guaraná Fantástica', Decimal('4504.36'), 11),
 (1, 67, 'Laughing Lumberjack Lager', Decimal('2396.80'), 12),
 (2, 63, 'Vegie-spread', Decimal('16701.10'), 1),
 (2, 61, "Sirop d'érable", Decimal('14352.60'), 2),
 (2, 65, 'Louisiana Fiery Hot Pepper Sauce', Decimal('13869.89'), 3),
 (2, 8, 'Northwoods Cranberry Sauce', Decimal('12772.00'), 4),
 (2, 44, 'Gula Malacca', Decimal('9915.95'), 5),
 (2, 77, 'Original Frankfurter grüne Soße', Decimal('9171.63'), 6),
 (2, 4, "Chef Anton's Cajun Seasoning", Decimal('8567.90'), 7),
 (2, 6, "Grandma's Boysenberry Spread", Decimal('7137.00'), 8)]

Using two `CTE` to check for the top products based on sales for each category:

In [17]:
%%sql

WITH top_products_table AS (
SELECT p.category_id,c.category_name,od.product_id,p.product_name,
       ROUND(CAST(SUM(od.quantity*od.unit_price*(1-od.discount)) AS numeric),2) AS product_sales
  FROM order_details AS od
  JOIN products AS p
 USING (product_id)
  JOIN categories AS c
 USING (category_id)
 GROUP BY od.product_id,p.product_name,p.category_id,c.category_name
 ORDER BY category_id
    
),

ranked_products_table AS (
SELECT category_id,product_id,product_name,product_sales, 
       DENSE_RANK() OVER (
       PARTITION BY category_id
       ORDER BY product_sales DESC
       ) AS product_ranking
  FROM top_products_table
)

SELECT category_id, product_id,product_name AS top_product,product_sales
  FROM ranked_products_table
 WHERE product_ranking = 1

 * postgresql://postgres:***@localhost:5432/northwind
8 rows affected.


[(1, 38, 'Côte de Blaye', Decimal('141396.74')),
 (2, 63, 'Vegie-spread', Decimal('16701.10')),
 (3, 62, 'Tarte au sucre', Decimal('47234.97')),
 (4, 59, 'Raclette Courdavault', Decimal('71155.70')),
 (5, 56, 'Gnocchi di nonna Alice', Decimal('42593.06')),
 (6, 29, 'Thüringer Rostbratwurst', Decimal('80368.67')),
 (7, 51, 'Manjimup Dried Apples', Decimal('41819.65')),
 (8, 18, 'Carnarvon Tigers', Decimal('29171.87'))]

The above table shows the top products based on sales for each category. These are the top priorities for marketing and restocking, especially Cote de Blaye